File: scripts/Qual Coding Example.R
# -------------------------------------------------------------------
# Author: Andrew Avitabile
# Purpose: General function to that uses ChatGPT to code things
# -------------------------------------------------------------------

# ---- Setup ----
rm(list = ls())
# You'll need to install the pacman package and these below:
pacman::p_load("httr", "jsonlite", "tidyverse", "progress")

# -------------------------------------------------------------------
# Function: call_chatgpt_batch()
# -------------------------------------------------------------------
# Description:
#   Takes a dataset and a text column, sends each unique text value to
#   the ChatGPT API using a user-defined prompt, and returns responses.
#
# Arguments:
#   data       - data frame containing text to classify
#   text_col   - column name (string) of text input
#   prompt_fn  - function that builds the prompt given one text value
#   model      - OpenAI model to use (e.g., "gpt-4o-mini", "gpt-3.5-turbo")
#   api_key    - your OpenAI API key (you'll need to get this by making an account at https://openai.com/. 
#                Try to keep this a secret, as it can be used to bill your account. Consider saving in .Renviron file)
#   start_item - index to resume from (useful for long lists)
#   batch_n    - number of items per batch (for progress tracking and in case the code breaks)
#
# Returns:
#   Original data joined with model responses
# -------------------------------------------------------------------

call_chatgpt_batch <- function(data,
                               text_col,
                               prompt_fn,
                               model = "gpt-4o-mini",
                               api_key = Sys.getenv("OPENAI_API_KEY"),
                               start_item = 1,
                               batch_n = 100) {
  
  # Get unique text values
  texts <- unique(data[[text_col]])
  total <- length(texts)
  
  if (start_item > 1) {
    texts <- texts[start_item:total]
    cat("Resuming from item", start_item, "| Remaining:", length(texts), "\n")
  }
  
  cat("Total unique texts:", total, "| Processing in batches of", batch_n, "\n")
  
  # --- Helper function to call API for one item ---
  get_response <- function(text_value) {
    prompt <- prompt_fn(text_value)
    
    res <- httr::POST(
      url = "https://api.openai.com/v1/chat/completions",
      httr::add_headers(Authorization = paste("Bearer", api_key)),
      httr::content_type_json(),
      body = jsonlite::toJSON(list(
        model = model,
        messages = list(list(role = "user", content = prompt)),
        temperature = 0
      ), auto_unbox = TRUE)
    )
    
    result <- httr::content(res, as = "parsed")
    output <- result$choices[[1]]$message$content
    return(trimws(output))
  }
  
  # --- Process all items ---
  all_results <- tibble()
  
  for (batch_num in 1:ceiling(length(texts) / batch_n)) {
    start_idx <- (batch_num - 1) * batch_n + 1
    end_idx <- min(batch_num * batch_n, length(texts))
    batch_texts <- texts[start_idx:end_idx]
    
    cat("\nProcessing batch", batch_num, "(", length(batch_texts), "items)\n")
    
    pb <- progress_bar$new(
      format = "[:bar] :current/:total (:percent)",
      total = length(batch_texts), clear = FALSE, width = 60
    )
    
    batch_responses <- sapply(batch_texts, function(x) {
      out <- tryCatch(get_response(x), error = function(e) NA)
      pb$tick()
      return(out)
    }, USE.NAMES = TRUE)
    
    batch_df <- tibble(!!text_col := names(batch_responses),
                       model_response = unname(batch_responses))
    
    all_results <- bind_rows(all_results, batch_df)
  }
  
  # Join back with original data
  data_joined <- left_join(data, all_results, by = text_col)
  
  cat("\n✅ All batches complete.\n")
  return(data_joined)
}

# -------------------------------------------------------------------
# Toy Example (ready to run)
# -------------------------------------------------------------------
toy_data <- tibble(item = c("Apple", "Carrot", "Strawberry", "Broccoli"))

toy_prompt <- function(food) {
  paste0("Is '", food, "' a fruit or a vegetable? Respond with only one word: 'fruit' or 'vegetable'.")
}

toy_results <- call_chatgpt_batch(
  data = toy_data,
  text_col = "item",
  prompt_fn = toy_prompt,
  model = "gpt-4o-mini",
  batch_n = 10
)

print(toy_results)


# -------------------------------------------------------------------
# Toy Example (ready to run)
# -------------------------------------------------------------------
# Example data: 3 policies
toy_data <- tibble(item = c(
  # 1. Has explicit number
  "Paternity Leave The school recognizes the importance of supporting employees in balancing their professional responsibilities with the arrival of a new child. Eligible employees are entitled to up to four weeks of paternity leave following the birth, adoption, or placement of a child.",
  
  # 2. Mentions leave but no number
  "The school offers paternity leave to support new parents in caring for their child and adjusting to family routines. Employees should contact HR for eligibility and details.",
  
  # 3. No mention of paternity leave
  "The school encourages professional development and provides up to five days of paid leave per year for attending educational workshops."
))

# Prompt: detect mention + extract number (or 0/NA)
toy_prompt <- function(policy) {
  paste0(
    "You are a careful qualitative researcher. ",
    "Read the policy text below and do the following:\n",
    "1. If the policy mentions paternity leave, extract the number of days it provides.\n",
    "2. If it mentions paternity leave but no number is stated, return 0.\n",
    "3. If it does not mention paternity leave at all, return NA.\n",
    "Convert weeks or months to days (7 days per week, 30 days per month).\n",
    "Respond ONLY with a single number (e.g., 28, 0, or NA).\n\n",
    policy
  )
}

# Call the general function
toy_results <- call_chatgpt_batch(
  data = toy_data,
  text_col = "item",
  prompt_fn = toy_prompt,
  model = "gpt-4o-mini",
  batch_n = 10
)

print(toy_results)

File: scripts/Qual Coding Example.py
# Author: Andrew Avitabile
# Purpose: This script analyzes teacher feedback using ChatGPT (GPT-4o-mini) to identify areas for improvementin teaching practice. 
# It processes feedback stored in an Excel file and saves the results.

# ============== SETUP INSTRUCTIONS ==============
# Before running this script, you need to:
# 1. Install required packages by running these commands in your terminal:
#    pip install pandas openai python-dotenv
#
# 2. Create a .env file containing your OpenAI API key like this:
#    OPENAI_API_KEY=your-key-here
#
# 3. Prepare your Excel file with at least these columns:
#    - text: contains the feedback to be analyzed
#    Note: Other columns will be preserved in the output

# Import required Python packages
import os                   # For working with file paths
import pandas as pd         # For handling Excel files and data processing
from openai import OpenAI   # For communicating with ChatGPT
from dotenv import load_dotenv  # For loading API key from .env file

# This class contains all the logic for analyzing feedback
class SimpleFeedbackAnalyzer:
    # This is a special method that runs when you create a new analyzer. It's like setting up your workspace before you start working.
    # self refers to the specific instance of the analyzer you're creating. It's like saying "this particular analyzer"
    # The other methods (create_prompt and get_area_for_improvement) are tools or functions that belong to this analyzer
    def __init__(self, env_path): 
        # This is information for future people who use the analyzer.
        """
        Sets up the feedback analyzer by:
        1. Loading the OpenAI API key
        2. Setting up the connection to OpenAI
        3. Defining the possible teaching skills to look for
        
        Args:
            env_path (str): Path to your .env file containing the OpenAI API key
        """
        # Load the API key from a specified .env file (env_path)
        load_dotenv(env_path)
        
        # Set up the connection to OpenAI's API
        self.client = OpenAI(
            api_key=os.getenv("OPENAI_API_KEY")  # Get API key from environment variables (i.e., you need OPENAI_API_KEY in your .env file)
        )
        
        # List of teaching skills that ChatGPT will look for in the feedback
        # You can modify this list to look for different skills or change what you're looking for in the text
        self.teaching_skills = [
            "Classroom Management",
            "Lesson Planning",
            "Differentiation",
            "Assessment and Feedback",
            "Student Engagement",
            "Student Comprehension",
            "Communication"
        ]

    # Now we're defining a prompt to use for each row of text. 
    # Note that it requires a string of text (e.g., the feedback) 
    # Also note that it returns a string with the feedback surrounded by the prmopt
    def create_prompt(self, feedback_text):
        """
        Creates the instruction text (prompt) that will be sent to ChatGPT.
        
        Args:
            feedback_text (str): The feedback text to analyze
            
        Returns:
            str: The complete prompt for ChatGPT
        """
        # Convert the list of skills into a comma-separated string
        skills_list = ", ".join(self.teaching_skills)
        
        # Create the prompt that tells ChatGPT exactly what to do. This is where you may need to beg a bit.
        return f"""Analyze this feedback given to a pre-service teacher and identify the main area that needs improvement.
        
Feedback text:
{feedback_text}

Respond with ONLY ONE of these options: {skills_list}, "other", "none", or "multiple".

Rules:
- Choose "multiple" if there are several equally emphasized areas for improvement
- Choose "none" if no specific area for improvement is mentioned
- Choose "other" if the area for improvement doesn't match any of the listed skills
- Otherwise, choose the most prominent teaching skill that needs improvement"""

    def get_area_for_improvement(self, feedback_text):
        """
        Sends the feedback text to ChatGPT and gets back the area for improvement.
        
        Args:
            feedback_text (str): The feedback text to analyze
            
        Returns:
            str: The identified area for improvement
        """
        # Skip empty text to avoid errors
        if not feedback_text or len(feedback_text.strip()) < 0:
            return "none"

        try:
            # Send the request to ChatGPT
            response = self.client.chat.completions.create(
                model="gpt-4o-mini",  # The GPT model to use
                messages=[
                    # Tell ChatGPT its role
                    {
                        "role": "system",
                        "content": "You are an expert at analyzing teacher feedback. Respond with only the area that needs improvement."
                    },
                    # Provide the actual task and feedback
                    {
                        "role": "user",
                        "content": self.create_prompt(feedback_text)
                    }
                ],
                temperature=1  # Lower temperature means more consistent responses. If temperature is 0 you should get the same responses every time, but ChatGPT may be more risk-averse.
            )

            # Extract the response text from ChatGPT
            area = response.choices[0].message.content.strip()

            # Make sure the response is one of our expected values
            if area not in self.teaching_skills + ["other", "none", "multiple"]:
                area = "other"

            return area

        except Exception as e:
            # If anything goes wrong, print the error and return "none"
            print(f"Error analyzing feedback: {e}")
            return "none"

# Now we setup another function that will import our data from Excel and read our .env file
def process_excel_file(input_file, output_file, env_path):
    """
    Main function that:
    1. Reads the Excel file
    2. Processes each piece of feedback
    3. Saves the results to a new Excel file
    
    Args:
        input_file (str): Path to your input Excel file
        env_path (str): Path to your .env file with API key
    """
    try:
        # Load the Excel file into a pandas DataFrame
        print(f"Reading data from {input_file}...")
        df = pd.read_excel(input_file)
        
        # Create an instance of our feedback analyzer (i.e., we can run the SimpleFeedbackAnalyzer defined above using the word "analyzer")
        analyzer = SimpleFeedbackAnalyzer(env_path)
        
        # Add a new column to store our results
        df['area_for_improvement'] = 'none'
        
        # Process each row in the Excel file
        print("Analyzing feedback texts...")
        for index, row in df.iterrows():
            # Get the feedback text from the current row (note the importance of the text variable being present in the data)
            feedback_text = row['text']
            
            # Get the area for improvement from ChatGPT
            result = analyzer.get_area_for_improvement(feedback_text)
            
            # Save the result in our DataFrame. This is saved in our dataframe (df) at rownumber index and in column 'area_for_improvement'
            df.at[index, 'area_for_improvement'] = result
            
            # Show progress update every 10 rows. This is nice if you're running a lot of code.
            if (index + 1) % 10 == 0:
                print(f"Processed {index + 1} of {len(df)} rows")
        
        # Save the results to a new Excel file
        df.to_excel(output_file, index=False)
        print(f"\nAnalysis complete! Results saved to: {output_file}")
        
    except Exception as e:
        print(f"Error processing file: {e}")

def main():
    """
    Entry point of the script. Sets up file paths and starts the analysis.
    """
    # Specify where your files are located
    # IMPORTANT: Update these paths to match your computer!
    env_path = r"C:\Users\Andre\Dropbox\ChatGPT Qual Example\scripts\.env"
    input_file = r"C:\Users\Andre\Dropbox\ChatGPT Qual Example\data\Example Data.xlsx"
    output_file = r"C:\Users\Andre\Dropbox\ChatGPT Qual Example\output\Example Data - Coded.xlsx"
    
    # Start processing the Excel file
    process_excel_file(input_file, output_file, env_path)

# This is the standard way to run a Python script
if __name__ == "__main__":
    main()